### ISSUE TITLE
Cannot search device by Custom Values (in API)

### ISSUE THREAD

**Issue by dw-sh-book** [samilliken/openDCIM#773]
Hi opendcim team,

It is so handy that we can see Custom Values in the api

```
curl -s -H -H "UserID:xxx" -H "APIKey:xxx" ''http://opendcim.org/demo/api/v1/device?Label=SW-AD01' | python -m json.tool | grep status
            "status": "Live"`
```

However, it would be better to support search by Custom Values as well, like this

```
curl  -H "UserID:xxx" -H "APIKey:xxx" 'http://opendcim.org/demo/api/v1/device?Label=SW-AD01&status=Live'
```

Could you kindly consider this ?

**Comment by wilpig** [comment]
The back end isn't setup to allow that

**Comment by dw-sh-book** [comment]
I see,  the Custom Attribute is in different table, it might need some join search and custom code to make it.

If api search by Custom Attribute is possible, it will make custom Attribute much useful, so that we can rely on opendcim as a configuration item  center.

Could you kindly consider this?

**Comment by FliesLikeABrick** [comment]
@samilliken  would you accept a 3rd party patch to add this support to 4.3.1?

**Comment by samilliken** [comment]
Yep, as long as you code no worse than I do.  LOL

Scott

**Comment by FliesLikeABrick** [comment]
@samilliken  any input on how we refer to custom values in the URL as a search criteria?  For example, some of ours have non-alphanum characters in them that might not be the best for using as a key in the URL.  With your input on that, there is a good probability we will work on a patch for this that I would like to share upstream once it is ready.

**Comment by FliesLikeABrick** [comment]
To clarify, I mean the names of custom fields, not the values

**Comment by samilliken** [comment]
Might want to check with Wilpig since he's been working on merging the data
model such that when you ask for $device variables, you get the base set
plus any custom attributes that are applied to it.  He might already be
thinking down the path of how to check for custom values.  I know we've
done a hack in the past that will search ALL custom value fields for the
given value, but that's not so handy if you expend the model a good bit.

Scott

**Comment by wilpig** [comment]
The models have been merged, basically we'd just need to alter the search method  in devices.php to know about custom attributes.  Since we pass in a partial model when we search we just need to specify in that function how to search for the custom attributes.

My thought is that we let the search that is functional today work, then perform a second search for the custom attributes that are present, then merge the two result sets, or run an intersection of them.  That part needs to be hashed out.  The main thing is it needs to go into the search function in general.

**Comment by wilpig** [comment]
Now that I look at that I may have broken searching when I merged in the model.

**Comment by FliesLikeABrick** [comment]
@wilpig  so what can we do to help/what work remains - or is this underway?  We have a project which would benefit from this ASAP, so I'm hoping to get it working here within a week or so -- but without duplicating your ongoing effort.

**Comment by Mustansir94** [comment]
It would be beneficial to have a search that queries the network devices connected to a console device. If somehow we could create a custom attribute that returns this information, it would be greatly useful!

**Comment by samilliken** [comment]
This was closed as part of #777 which added the ability in both the UI and API to search by Custom Attribute names (so not having to know the key value from the db that the name relates to).
